packages/application-extension/src/index.tsx

```diff
@@ -429,38 +449,196 @@ const sidebar: JupyterFrontEndPlugin<void> = {
+    const test = (node: HTMLElement) =>
+      node['title'] && !!node['title'].match(pathRe);
+    const node = app.contextMenuHitTest(test);
+
+    if (!node) {
+      // Fall back to active doc widget if path cannot be obtained from event.
+      return labShell.currentWidget;
+    }
+    const pathMatch = node['title'].match(pathRe);
+    return docManager.findWidget(pathMatch[1]);
+  };
+
+  // Closes an array of widgets.
+  const closeWidgets = (widgets: Array<Widget>): void => {
+    widgets.forEach(widget => widget.close());
+  };
+
+  // Find the tab area for a widget within a specific dock area.
+  const findTab = (
+    area: DockLayout.AreaConfig,
+    widget: Widget
+  ): DockLayout.ITabAreaConfig | null => {
+    switch (area.type) {
+      case 'split-area':
+        const iterator = iter(area.children);
+        let tab: DockLayout.ITabAreaConfig | null = null;
+        let value: DockLayout.AreaConfig | null = null;
+        do {
+          value = iterator.next();
+          if (value) {
+            tab = findTab(value, widget);
+          }
+        } while (!tab && value);
+        return tab;
+      case 'tab-area':
+        const { id } = widget;
+        return area.widgets.some(widget => widget.id === id) ? area : null;
+      default:
+        return null;
+    }
+  };
+
+  // Find the tab area for a widget within the main dock area.
+  const tabAreaFor = (widget: Widget): DockLayout.ITabAreaConfig | null => {
+    const { mainArea } = labShell.saveLayout();
+    if (mainArea.mode !== 'multiple-document') {
+      return null;
+    }
+    let area = mainArea.dock.main;
+    if (!area) {
+      return null;
+    }
+    return findTab(area, widget);
+  };
+
+  // Returns an array of all widgets to the right of a widget in a tab area.
+  const widgetsRightOf = (widget: Widget): Array<Widget> => {
+    const { id } = widget;
+    const tabArea = tabAreaFor(widget);
+    const widgets = tabArea ? tabArea.widgets || [] : [];
+    const index = widgets.findIndex(widget => widget.id === id);
+    if (index < 0) {
+      return [];
+    }
+    return widgets.slice(index + 1);
+  };
+
+  commands.addCommand(CommandIDs.activateNextTab, {
     label: 'Activate Next Tab',
     execute: () => {
       app.shell.activateNextTab();
     }
   });
-  palette.addItem({ command, category });
+  palette.addItem({ command: CommandIDs.activateNextTab, category });
 
-  command = CommandIDs.activatePreviousTab;
-  app.commands.addCommand(command, {
+  commands.addCommand(CommandIDs.activatePreviousTab, {
     label: 'Activate Previous Tab',
     execute: () => {
       app.shell.activatePreviousTab();
     }
   });
-  palette.addItem({ command, category });
+  palette.addItem({ command: CommandIDs.activatePreviousTab, category });
+
+  commands.addCommand(CommandIDs.close, {
+    label: () => {
+      // Returns the file type for a widget.
+      function fileType(widget: Widget, docManager: IDocumentManager): string {
```

Comment:
All this logic is making me think that the docmanager is in fact a better place for these.